docs/docs/adrs/adr-011-improving-test-confidence.md

```diff
@@ -0,0 +1,208 @@
+
+Examples of tests made easier with CometMock are listed below:
+- regression tests
+- non-determinism tests
+- upgrade tests
+- state-breaking changes
+
+With CometMock, the **matrix test** approach can also be used. The image below illustrates a CometMock setup that can be used to discover non-deterministic behavior and state-breaking changes.
+![e2e matrix tests](../../figures/cometmock_matrix_test.png)
+
+This table explains which versions are tested against each other for the same set of test traces:
+* ✅ marks a passing test
+* ❌ marks a failing test
+
+| **SCENARIO**                    | **start chain** | **add key** | **delegate** | **undelegate** | **redelegate** | **downtime** | **equivocation** | **stop chain** |
+|---------------------------------|-----------------|-------------|--------------|----------------|----------------|--------------|------------------|----------------|
+| **v3 provi + v3 consu**         | ✅               | ✅           | ✅            | ✅              | ✅              | ✅            | ✅                | ✅              |
+| **main provi + main consu**     | ✅               | ✅           | ✅            | ✅              | ✅              | ✅            | ✅                | ✅              |
+| **commit provi + commit consu** | ✅               | ❌           | ✅            | ❌              | ✅              | ✅            | ❌                | ❌              |
+
+
+
+Briefly; multiple versions of the application are run against the same CometMock instance and any deviations in app behavior would result in `app hash` errors (the apps would be in different states after performing the same set of actions).
+
+#### 3. Introduce innovative testing approaches
+
+When discussing e2e testing, some very important patterns emerge - especially if test traces are used instead of ad-hoc tests written by hand.
+
+We see a unique opportunity to clearly identify concerns and modularize the testing architecture. 
+
+The e2e testing frameworks can be split into a **pipeline consisting of 3 parts: model, driver and harness**.
+
+##### Model
+
+Model is the part of the system that can emulate the behavior of the system under test.
+Ideally, it is very close to the specification and is written in a specification language such as quint, TLA+ or similar.
+One of the purposes of the model is that it can be used to generate test traces. 
+
+
+##### Driver
+
+The purpose of the driver is to accept test traces (generated by the model or written by hand), process them and provide inputs to the next part of the pipeline.
+
+Basically, the driver sits between the model and the actual infrastructure on which the test traces are being executed on.
+
+##### Harness
+
+Harness is the infrastructure layer of the pipeline that accepts inputs from the driver.
+
+There can be multiple harnesses as long as they can perform 3 things:
+* bootstrap a test execution environment (local, docker, k8s…)
```

Comment:
Something that is not clear to me is on how can we guarantee that the driver can execute the _exact same_ trace as it was given by the model. For example, consider we use docker containers, and the model-generated trace that we have is a sequence of events like: "chain `A` sends message `m`, chain `B` sends message `m'`, chain `A` receives `m'`, ...` then to run this _exact_ trace we would need to enforce how/when the relayer relays packets or coordinate when chains `A` and `B` take steps. Is this something we can (easily) do?

Reply:
Relayers can be controlled in a relatively fine-grained manner - you can in general make them clear packets on a given channel, then stop clearing more packets. What you described is easily possible in principle - just clear the channel that m' was sent on first.

Coordinating when exactly certain steps are taken is more difficult, it's e.g. possible with CometMock.

But generally, we often do not assume these very strong orderings. For example, assume we want to delegate on `A`, this triggers a message m to be sent from `A` to `B`, we want to first submit a `tx` to `B`, then `B` should receive `m`, then we want to submit another `tx'` to `B`. This is easily possible - we submit the delegation to `A`, wait until it is included, submit `tx` to `B`, wait until it is included, have the relayer relay `m` to `B`, wait until m was included, submit `tx'` to `B`, wait until it is included. Only the relative order of the events matters, we don't need precise coordination between block numbers on either chain.

Does that cover the kinds of examples you were thinking of?

Reply:
Yup. Thank you very much @p-offtermatt !